From 317c52d9f02e66c160d352b61c838c3ca0ae1616 Mon Sep 17 00:00:00 2001
From: "Roy, UjjalX" <uroy@maxlinear.com>
Date: Wed, 22 Sep 2021 19:11:00 +0300
Subject: [PATCH] Merge pull request #3540 in SW_UGW/linux from
 feature/UGW_SW-61220-mcast-fallback-helper-learning-8.5.2 to 8.5.2

* commit '5c0d0e4e1f83bb46f10c4de086eb76f770d26c2a':
  UGW_SW-61220 : Fallback mode helper learning
---
 net/bridge/br_forward.c | 16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/net/bridge/br_forward.c b/net/bridge/br_forward.c
index 7a4c86c10b01..c491ef00d349 100644
--- a/net/bridge/br_forward.c
+++ b/net/bridge/br_forward.c
@@ -47,7 +47,7 @@ static void mcast_helper_learning_hook(struct sk_buff *skb)
 		}
 	}
 
-#ifdef CONFIG_MCAST_SNOOPING
+#if IS_ENABLED(CONFIG_MCAST_SNOOPING)
 	/* Hook the UDP multicast to mcast helper */
 	if (!bridge_lanserver_hook)
 		return;
@@ -213,7 +213,7 @@ static struct net_bridge_port *maybe_deliver(
 	if (!prev)
 		goto out;
 
-#ifdef CONFIG_MCAST_SNOOPING
+#if IS_ENABLED(CONFIG_MCAST_SNOOPING)
 	if ((bridge_igmp_snooping || bridge_mld_snooping) &&
 	    is_multicast_ether_addr(eth_hdr(skb)->h_dest) &&
 	    (br_selective_flood(prev, skb) == 0)) {
@@ -238,6 +238,14 @@ void br_flood(struct net_bridge *br, struct sk_buff *skb,
 	struct net_bridge_port *prev = NULL;
 	struct net_bridge_port *p;
 
+#if IS_ENABLED(CONFIG_MCAST_SNOOPING)
+#if IS_ENABLED(CONFIG_MCAST_HELPER)
+	if (is_multicast_ether_addr(eth_hdr(skb)->h_dest) &&
+	    !is_broadcast_ether_addr(eth_hdr(skb)->h_dest))
+		mcast_helper_learning_hook(skb);
+#endif
+#endif
+
 	list_for_each_entry_rcu(p, &br->port_list, list) {
 		/* Do not flood unicast traffic to ports that turn it off */
 		if (pkt_type == BR_PKT_UNICAST && !(p->flags & BR_FLOOD))
@@ -266,7 +274,7 @@ void br_flood(struct net_bridge *br, struct sk_buff *skb,
 		goto out;
 
 	if (local_rcv) {
-#ifdef CONFIG_MCAST_SNOOPING
+#if IS_ENABLED(CONFIG_MCAST_SNOOPING)
 	if ((bridge_igmp_snooping || bridge_mld_snooping) &&
 	    is_multicast_ether_addr(eth_hdr(skb)->h_dest) &&
 	    (br_selective_flood(prev, skb) == 0)) {}
@@ -274,7 +282,7 @@ void br_flood(struct net_bridge *br, struct sk_buff *skb,
 #endif
 		deliver_clone(prev, skb, local_orig);
 	} else {
-#ifdef CONFIG_MCAST_SNOOPING
+#if IS_ENABLED(CONFIG_MCAST_SNOOPING)
 	if ((bridge_igmp_snooping || bridge_mld_snooping) &&
 	    is_multicast_ether_addr(eth_hdr(skb)->h_dest) &&
 	    (br_selective_flood(prev, skb) == 0))
